### PR TITLE
dynamodocstore: separate query planning from execution

### PR DESCRIPTION
**Pull request by jba**
Before the first RPC of a query, create the ScanInput or QueryInput
for it.  Reuse that input for subsequent RPCs.

Also, redo the limit logic since the previous version had a couple of
problems.

### CODE REVIEW

File: internal/docstore/dynamodocstore/query.go [hunk 1]
```diff
@@ -68,41 +71,55 @@ func (c *collection) runQueryOrScan(ctx context.Context, q *driver.Query, startA
...
+
+type queryRunner struct {
+	c       *collection
+	scanIn  *dynamodb.ScanInput
+	queryIn *dynamodb.QueryInput
+}
+
+func (qr *queryRunner) run(ctx context.Context, startAfter avmap) (items []avmap, last avmap, err error) {
+	if qr.scanIn != nil {
+		qr.scanIn.ExclusiveStartKey = last
```
Comment (shantuo) [inline internal/docstore/dynamodocstore/query.go]: Should `last` be `startAfter`?

File: internal/docstore/dynamodocstore/query.go [hunk 2]
```diff
@@ -68,41 +71,55 @@ func (c *collection) runQueryOrScan(ctx context.Context, q *driver.Query, startA
...
+		out, err := qr.c.db.ScanWithContext(ctx, qr.scanIn)
+		if err != nil {
+			return nil, nil, err
+		}
+		return out.Items, out.LastEvaluatedKey, nil
+	}
+	qr.queryIn.ExclusiveStartKey = last
```
Comment (shantuo) [inline internal/docstore/dynamodocstore/query.go]: Same here.

File: internal/docstore/dynamodocstore/query.go (hunk 1 above)
Comment (jba) [reply internal/docstore/dynamodocstore/query.go]: Good catch. Fixed.

File: internal/docstore/dynamodocstore/query.go (hunk 2 above)
Comment (jba) [reply internal/docstore/dynamodocstore/query.go]: ditto

**Comment by jba** [comment]
PTAL

**Pull request merged**